diff -rup lloyd-yajl-9c15d72.orig/src/CMakeLists.txt lloyd-yajl-9c15d72/src/CMakeLists.txt
--- lloyd-yajl-9c15d72.orig/src/CMakeLists.txt	2010-01-08 09:05:11.000000000 -0800
+++ lloyd-yajl-9c15d72/src/CMakeLists.txt	2013-08-16 16:24:36.614552143 -0700
@@ -42,7 +42,7 @@ SET (PUB_HDRS api/yajl_parse.h api/yajl_
 ADD_DEFINITIONS(-DYAJL_BUILD)
 
 # set up some paths
-SET (libDir ${CMAKE_CURRENT_BINARY_DIR}/../${YAJL_DIST_NAME}/lib)
+SET (libDir ${CMAKE_CURRENT_BINARY_DIR}/../${YAJL_DIST_NAME}/${LIB_INSTALL_DIR})
 SET (incDir ${CMAKE_CURRENT_BINARY_DIR}/../${YAJL_DIST_NAME}/include/yajl)
 
 # set the output path for libraries
@@ -77,8 +77,8 @@ ENDFOREACH (header ${PUB_HDRS})
 INCLUDE_DIRECTORIES(${incDir}/..)
 
 IF(NOT WIN32)
-  INSTALL(TARGETS yajl LIBRARY DESTINATION lib)
-  INSTALL(TARGETS yajl_s ARCHIVE DESTINATION lib)
+  INSTALL(TARGETS yajl LIBRARY DESTINATION ${LIB_INSTALL_DIR})
+  INSTALL(TARGETS yajl_s ARCHIVE DESTINATION ${LIB_INSTALL_DIR})
   INSTALL(DIRECTORY api/ DESTINATION include/yajl
           FILES_MATCHING PATTERN "*.h")
 ENDIF()
